File: jdbc2/org/softlang/features/Depth.java
package org.softlang.features;

import org.softlang.company.*;

/**
 * Get the company depth
 * 
 */
public class Depth {

	public static int depth(Company company) {
		int maxDepth = 0;
		for (Department dept : company.getDepts())
			maxDepth = 1 + Math.max(maxDepth, depth(dept));
		return maxDepth;
	}

	public static int depth(Department dept) {
		int maxDepth = 0;
		for (Department subDepartment : dept.getSubDepartments())
			maxDepth = 1 + Math.max(maxDepth, depth(subDepartment));
		return maxDepth;
	}

}

File: contributions/haskellVariation/Depth.hs
module Depth where

import Company

depth :: Company -> Int
depth (Company _ ds) = max' (map depth' ds)
  where
    max' = foldr max 0
    depth' :: Department -> Int
    depth' (Department _ _ sus) = 1 + max' (map depth'' sus)
    depth'' :: SubUnit -> Int
    depth'' (EUnit _) = 0
    depth'' (DUnit d) = depth' d
